Added emulation for accesses with effective privelege mode = S

Wesdcv opened the pull request:
**Draft PR:**
For now in order for this to work on VisionFive2 board a modified version of linux is required. 
It also breaks `sandbox` test CI.

Wesdcv commented:
Before making additional changes, on this branch, i wanted to check a couple of things with @CharlyCst. 
1. How should we go about validating the virtual addresses? From my perspective, there are three edge cases that we need to think about: access to a general memory (which is permitted), access to a Miralis-PMP-protected memory region (which is denied), and access to device-PMP-memory (which is not clear if it's valid via a virt address). 
2. Empirically, it seems like there weren't any traps encountered on the board during execution of `handle_virtual_load_store()`. 
https://github.com/CharlyCst/miralis/blob/36abbd7cf324f757034d70f1aa75db5bece0de17/src/arch/metal.rs#L532 However, should there be support for such a possibility in the code?
3. I was just wondering if we should keep the case for ecall from OS with `fid=20` (that's a random number for MIP patching) in the codebase or if we should look at the underlying issue in Linux to see if we can get rid of the need for ecall?
https://github.com/CharlyCst/miralis/blob/36abbd7cf324f757034d70f1aa75db5bece0de17/src/virt.rs#L622-L638

CharlyCst commented:
To give a quick answer (we can discuss more in details next week or by following up here):

1. My understanding is that when `MPRV=1` then the PMP take effect as if S or U mode was executing. We should double check what the spec says around that, but it seems to contradict what you observed? If PMP are taken into account we only need to worry about (3), and we already have a function to check if we touch a device (for now we can panic if that's the case, I don't think we will need it for Linux).
2. Yes, we should handle that case and inject back an exception. OpenSBI handles it too by injecting the error into Linux.
3. In the end we will need to run unmodified Linux, so we don't want to keep the hack around. We will need to investigate into this issue…

Wesdcv commented:
**Changes to the draft state:**
- Removed the interrupts patch from this branch
- Fixed sandbox issue (PMP with `idx=0` is disabled now instead of allowing `RWX`, due to check priority it was allowing all the accesses)
- Overall refactored the code a bit

Tested with Linux payload on `Spike` - works fine. The PR is not yet ready to be merged to my mind, but i would like to get some feedback now.

Wesdcv commented:
> If PMP are taken into account we only need to worry about (3), and we already have a function to check if we touch a device (for now we can panic if that's the case, I don't think we will need it for Linux).

Yes, we check if the mtval corresponds to any memory-mapped physical address, i was asking more about if access like this could ever be a case:
```
                                                                                         
                                        ┌───────┐                                        
                                       ┌───────┐│                                        
                                       │  PT   ││                                        
                        vMPRV=1        │       ││                                        
                 ┌────────────────────►│       │├────────────────────────┐               
                 │                     │       ││                        │               
                 │                     │       │┘                        │               
                 │                     └───────┘                         │               
   ┌─────────────┴───────────┐                             ┌─────────────▼───────────┐   
   │                         │                             │                         │   
   │  ld a0, 0(0xFF...F8..)  │                             │  ld a0, 0(0x2000_0000)  │   
   │                         │                             │                         │   
   └─────────────────────────┘                             └─────────────────────────┘   
```                                                                                         
If such an access could happen, i won't be caught by
https://github.com/CharlyCst/miralis/blob/9d842700c3b9ec0f637eb2110ebb6f183f79b631/src/virt.rs#L499
and subsequently will go through direct emulation instead of a device access interface. 

I guess this question falls into the same basket as protecting Miralis's memory from access with MPRV=1 (which means we should somehow validate the virtual address).

CharlyCst commented:
Yes I think this should be caught by PMPs too... We could potentially write a test to ensure this is the case (see if the PMPs indeed prevents from reading Miralis memory or devices using MPRV).

Thanks for the PR! I'll review it when I get some time, most likely this Friday.
As this is a non-trivial feature, would you mind writing a description of how the emulation works? You can add that as a new section to the doc in `./docs/architecture.md`.

CharlyCst commented on src/virt.rs:
```diff
@@ -508,15 +506,27 @@ impl VirtContext {
                     );
                     self.handle_device_access_fault(&instr, &device);
+                } else if (self.csr.mstatus & mstatus::MPRV_FILTER) >> mstatus::MPRV_OFFSET == 1 {
+                    // Can an access to device be legally requested with a virtual address?
```
I think so yes. Is the address in `trap_info.mtval` a virtual address in this case?

Wesdcv replied:
Yes, one example would be `ff60000001d14076`.
Then we should somehow check if the virtual address corresponds to any device, and use device handler if that's the case (I don't have any ideas on how to do that though)

CharlyCst reviewed (commented):
Great work on this part :)

Overall looks good to me, just some small questions and nits to fix ^^
I think this part needs a bit of documentation, could you add a new section in the `docs/architecture.md` file to explain how we emulate `MPRV == 1`?

We might also want to think about adding a small tests, something similar to the sandbox firmware that does a load with `MPRV == 1` to a well known value in memory (event without page tables that might surface some bugs if we break the feature later one).

Wesdcv commented on src/arch/metal.rs:
```diff
@@ -531,6 +533,131 @@ impl Architecture for MetalArch {
+                    // Unsure if a trap might happen here with a correct emulation, probably not?
+
+                    ctx.trap_info.mcause = cause;
+                    ctx.trap_info.mstatus = mstatus;
```
Not sure if I'm following this
This bit:
https://github.com/CharlyCst/miralis/blob/9d842700c3b9ec0f637eb2110ebb6f183f79b631/src/arch/metal.rs#L609-L618
would be executed only if there was another trap while emulating the load/store, in which case the regular trap handler would not be invoked, and the TrapInfo would not be refreshed. From my understanding, the correct emulation of this scenario would include passing the new TrapInfo with the old mepc, am I wrong? Or how else should this be handled by `emulate_jump_trap_handler`?

CharlyCst replied:
Yes we need to pass some of the info, I was just worried about leaking state from the physical registers into the virtual context, but it turns out that should be done here: https://github.com/CharlyCst/miralis/blob/4b7ac9096d50f5491b029dcc1755269e00163edd/src/virt.rs#L425-L439

And indeed some of the bits from `pmip` are filtered for instance.
So I think your code is fine with respect to putting the physical registers in `trap_info` :)

Wesdcv commented on src/arch/metal.rs:
```diff
@@ -531,6 +533,131 @@ impl Architecture for MetalArch {
+
+                    "csrc mstatus, {0}",
+                    "csrw mtvec, {1}",
+                    out(reg) _,
+                    out(reg) _,
+                    out("t1") cause,
+                    out("t2") mtval,
+                    out("t4") mstatus,
+                    out("t6") mip,
```
I mean we could read it, but it's going to be an address of Miralis - what would it be used for?

CharlyCst replied:
You should just tell the compiler that the register might get corrupted by the assembly snippet, otherwise you might get undefined behavior.

Adding something like `out("t5") _` will fix the issue

Wesdcv commented on src/virt.rs:
```diff
@@ -508,15 +506,27 @@ impl VirtContext {
                 }
             }
             MCause::InstrAccessFault => {
+                let instr = unsafe { Arch::get_raw_faulting_instr(&self.trap_info) };
+                let instr = decode(instr);
+                log::trace!("Instruction access fault: {:x?}", instr);
```
Well, currently decoded instruction is required for emulation purposes, is there is a workaround for that?
A bit uncertain about what's bothering you exactly (and why is this use case different from the others), but from my understanding `trap_info.mepc` should be filled by hardware, thus valid, and the question of access address validity kind of falls in the same bucket as miralis/devices protection

CharlyCst replied:
Yes it is needed when we emulate the instruction, but here this is inside a match arm where we do not emulate an instruction but simply inject a trap. So the `instr` variable is only used for printing in the above snippet.

The problem here is that we have no guarantee that the address holds a valid instruction, because the trap was an instruction access fault, not a trap due to a privileged instruction (which would imply that the instruction is valid).

Wesdcv replied:
Oh, sorry, my bad, didn't check the lines. In this case yes, you're right.

Wesdcv commented on src/arch/metal.rs:
```diff
@@ -531,6 +533,131 @@ impl Architecture for MetalArch {
+
+                    ctx.trap_info.mcause = cause;
+                    ctx.trap_info.mstatus = mstatus;
+                    ctx.trap_info.mtval = mtval;
```
That would be virtual address. In theory, the only way it could get anything other that this would be catching an `Illegal instruction` exception, which should be impossible.

Wesdcv commented:
Addressed most of the issues, what's left:

- Figuring out if `handle_virtual_load_store` requires any assumptions to be safe.
- Deciding if virtual address should be validated (even if PMP denies access properly, we still would probably like to handle access to a device with a virtual address somehow).

CharlyCst commented:
Ping me when you would like me to take another look :)

There is a "re-request review" button at the top right in the "Reviewers" section adds a "to-review" item in my notification list :)

CharlyCst reviewed (commented):
Looks good to me, just some minor fixes for the test, but beside that looks all good :)
Regarding the virtual address validation, my intuiting and understanding is that the PMP should catch if the access is invalid, but let's keep the TODO and write a test later on to verify that :)

Thanks for pushing this big feature through! It's a cool way of emulating MPRV :)

Wesdcv commented:
> Regarding the virtual address validation, my intuiting and understanding is that the PMP should catch if the access is invalid

That should indeed be the case for addresses that we're purely protecting (like Miralis's memory), but if we're using PMP as a mean of catching access attempt (like with devices), an access fault would be returned to firmware instead of a correct access emulation chain. That might be an undesired behavior.

CharlyCst reviewed (approved):
Awesome, looks all good to me :)
You did a great job on that one! Thanks for pushing it through!
